operator/src/test/java/org/keycloak/operator/testsuite/apiserver/ApiServerHelper.java

```diff
@@ -41,25 +41,30 @@ public void onAdd(StatefulSet obj) {
             }
 
             private void updateStatefulSet(StatefulSet obj) {
-                int replicas = obj.getSpec().getReplicas();
+                int specReplicas = obj.getSpec().getReplicas();
+                int statusReplicas = Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0);
                 String revision = Utils.hash(List.of(obj.getSpec())).substring(1);
-                if (!Objects.equals(Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0), replicas)
-                        || !Objects.equals(revision, obj.getStatus().getUpdateRevision())
-                        || !Objects.equals(obj.getStatus().getCurrentRevision(), obj.getStatus().getUpdateRevision())) {
+                String updateRevision = obj.getStatus().getUpdateRevision();
+
+                if (statusReplicas != specReplicas
+                        || !Objects.equals(revision, updateRevision)
+                        || !Objects.equals(obj.getStatus().getCurrentRevision(), updateRevision)) {
                     // generate intermediate rolling events
                     int actualReplicas;
-                    if (!Objects.equals(revision, obj.getStatus().getUpdateRevision())) {
+                    if (!Objects.equals(revision, updateRevision)) { // detected spec change, mimic rolling update
+                        // this is not fully accurate as it's rather recreate than rolling update,
+                        // but thanks to gradual scaling up it emits more events which is closer to the real behavior
                         actualReplicas = 0;
-                    } else if (replicas == 0) {
-                        actualReplicas = Math.max(replicas, Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0) - 1);
+                    } else if (specReplicas == 0 && statusReplicas > 0) { // probably recreate update requested, scaling down the deployment
```

Comment:
If specReplicas == 0 and statusReplicas == 0, we'll end up in the else case and scale up.

Reply:
I might be wrong but I do not think it'd scale up.

That is thanks to
https://github.com/shawkins/keycloak/blob/a430388622a95b799125ea809c368386842bf517/operator/src/test/java/org/keycloak/operator/testsuite/apiserver/ApiServerHelper.java#L49-L51
and
https://github.com/shawkins/keycloak/blob/a430388622a95b799125ea809c368386842bf517/operator/src/test/java/org/keycloak/operator/testsuite/apiserver/ApiServerHelper.java#L67-L69

Or am I missing something?

Reply:
Oh, we'd end up scaling up if you created a StatefulSet with initial 0 replicas. That's what you mean?

Reply:
Fixed that as well.

Reply:
It should also account for not incrementing past the actual desired number of replicas.